hudi-hadoop-mr/src/test/java/org/apache/hudi/hadoop/utils/TestHoodieArrayWritableAvroUtils.java

```diff
@@ -118,4 +124,282 @@ public void testCastOrderingField() {
+    Writable result = HoodieArrayWritableAvroUtils.rewritePrimaryType(bool, boolSchema, boolSchema);
+    assertSame(bool, result);
+  }
+
+  @Test
+  void testUnsupportedRewriteMapToIntThrows() {
+    Schema oldSchema = Schema.createMap(Schema.create(Schema.Type.STRING));
+    Schema newSchema = Schema.create(Schema.Type.INT);
+    assertThrows(HoodieAvroSchemaException.class, () ->
+        HoodieArrayWritableAvroUtils.rewritePrimaryType(new Text("foo"), oldSchema, newSchema));
+  }
+
+  @Test
+  void testRewriteIntToDecimalFixed() throws AvroSerdeException {
+    Schema fixedDecimal = LogicalTypes.decimal(8, 2).addToSchema(Schema.createFixed("dec", null, null, 6));
+    IntWritable oldWritable = new IntWritable(12345);
+    Writable result = HoodieArrayWritableAvroUtils.rewritePrimaryType(oldWritable, Schema.create(Schema.Type.INT), fixedDecimal);
+    assertInstanceOf(HiveDecimalWritable.class, result);
+    assertEquals(new BigDecimal("12345"), ((HiveDecimalWritable) result).getHiveDecimal().bigDecimalValue());
+    validateRewriteWithAvro(oldWritable, Schema.create(Schema.Type.INT), result, fixedDecimal);
+  }
+
+  @Test
+  void testRewriteDoubleToDecimalFixed() throws AvroSerdeException {
+    Schema fixedDecimal = LogicalTypes.decimal(10, 3).addToSchema(Schema.createFixed("dec", null, null, 8));
+    DoubleWritable oldWritable = new DoubleWritable(987.654);
+    Writable result = HoodieArrayWritableAvroUtils.rewritePrimaryType(oldWritable, Schema.create(Schema.Type.DOUBLE), fixedDecimal);
+    assertInstanceOf(HiveDecimalWritable.class, result);
+    assertEquals(new BigDecimal("987.654"), ((HiveDecimalWritable) result).getHiveDecimal().bigDecimalValue());
+    validateRewriteWithAvro(oldWritable, Schema.create(Schema.Type.DOUBLE), result, fixedDecimal);
+  }
+
+  @Test
+  void testRewriteDecimalBytesToFixed() throws AvroSerdeException {
+    Schema decimalSchema = LogicalTypes.decimal(6, 2).addToSchema(Schema.createFixed("dec", null, null, 6));
+    BigDecimal value = new BigDecimal("999.99");
+    byte[] unscaledBytes = value.unscaledValue().toByteArray();
+    BytesWritable oldWritable = new BytesWritable(unscaledBytes);
+    Writable result = HoodieArrayWritableAvroUtils.rewritePrimaryType(oldWritable, Schema.create(Schema.Type.BYTES), decimalSchema);
+    assertEquals(value, ((HiveDecimalWritable) result).getHiveDecimal().bigDecimalValue());
+    validateRewriteWithAvro(oldWritable, Schema.create(Schema.Type.BYTES), result, decimalSchema);
+  }
+
+  private void validateRewriteWithAvro(
+      Writable oldWritable,
+      Schema oldSchema,
+      Writable newWritable,
+      Schema newSchema
+  ) throws AvroSerdeException {
+    TypeInfo oldTypeInfo = HiveTypeUtils.generateTypeInfo(oldSchema, Collections.emptySet());
+    TypeInfo newTypeInfo = HiveTypeUtils.generateTypeInfo(newSchema, Collections.emptySet());
+
+    ObjectInspector oldOI = TypeInfoUtils.getStandardJavaObjectInspectorFromTypeInfo(oldTypeInfo);
```

Comment:
Although it may be verbose, my preference is to use ObjectInspector instead of `OI` in the variable and method names